hbase-server/src/main/java/org/apache/hadoop/hbase/master/snapshot/TakeSnapshotHandler.java

```diff
@@ -147,12 +152,16 @@ private TableDescriptor loadTableDescriptor() throws IOException {
   public TakeSnapshotHandler prepare() throws Exception {
     super.prepare();
     // after this, you should ensure to release this lock in case of exceptions
-    this.tableLock.acquire();
-    try {
-      this.htd = loadTableDescriptor(); // check that .tableinfo is present
-    } catch (Exception e) {
-      this.tableLock.release();
-      throw e;
+    if (this.tableLock.tryAcquire(this.lockAcquireTimeoutMs)) {
+      try {
+        this.htd = loadTableDescriptor(); // check that .tableinfo is present
+      } catch (Exception e) {
+        this.tableLock.release();
+        throw e;
+      }
+    } else {
+      LOG.error("Master lock could not be acquired in {} ms", lockAcquireTimeoutMs);
+      throw new IOException("Master lock could not be acquired");
```

Comment:
I don't think retrying makes much sense in this scenario, so I would prefer to throw a DoNotRetryIOException like HBaseSnapshotException or SnapshotCreationException. What do you think ?

Reply:
Actually, regardless of the Exception type here, it does eventually throw `SnapshotCreationException`:

```
    try {
      handler.prepare();
      this.executorService.submit(handler);
      this.snapshotHandlers.put(TableName.valueOf(snapshot.getTable()), handler);
    } catch (Exception e) {
      // cleanup the working directory by trying to delete it from the fs.
      Path workingDir = SnapshotDescriptionUtils.getWorkingSnapshotDir(snapshot, rootDir,
        master.getConfiguration());
      FileSystem workingDirFs = workingDir.getFileSystem(master.getConfiguration());
      try {
        if (!workingDirFs.delete(workingDir, true)) {
          LOG.error("Couldn't delete working directory (" + workingDir + " for snapshot:"
            + ClientSnapshotDescriptionUtils.toString(snapshot));
        }
      } catch (IOException e1) {
        LOG.error("Couldn't delete working directory (" + workingDir + " for snapshot:"
          + ClientSnapshotDescriptionUtils.toString(snapshot));
      }
      // fail the snapshot
      throw new SnapshotCreationException("Could not build snapshot handler", e,
        ProtobufUtil.createSnapshotDesc(snapshot));
    }
```

but sure i can still make it throw DoNotRetryIOException.

Reply:
Logs:

```
2024-01-03T15:17:21,015 ERROR [zk-snapshot {}] snapshot.TakeSnapshotHandler(163): Master lock could not be acquired in 0 ms
2024-01-03T15:17:21,016 DEBUG [PEWorker-3 {}] locking.LockProcedure(243): UNLOCKED pid=25, state=RUNNABLE, hasLock=true; org.apache.hadoop.hbase.master.locking.LockProcedure, tableName=TestSnapshotProcedureWithLockTimeout, type=EXCLUSIVE
2024-01-03T15:17:21,016 ERROR [zk-snapshot {}] procedure.TestSnapshotProcedureWithLockTimeout$1(103): zk snapshot failed
org.apache.hadoop.hbase.snapshot.SnapshotCreationException: Could not build snapshot handler
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.snapshotTable(SnapshotManager.java:678) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.snapshotEnabledTable(SnapshotManager.java:646) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.takeSnapshotInternal(SnapshotManager.java:762) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.takeSnapshot(SnapshotManager.java:707) ~[classes/:?]
	at org.apache.hadoop.hbase.master.procedure.TestSnapshotProcedureWithLockTimeout$1.run(TestSnapshotProcedureWithLockTimeout.java:101) ~[test-classes/:?]
Caused by: java.io.IOException: Master lock could not be acquired
	at org.apache.hadoop.hbase.master.snapshot.TakeSnapshotHandler.prepare(TakeSnapshotHandler.java:164) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.EnabledTableSnapshotHandler.prepare(EnabledTableSnapshotHandler.java:61) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.EnabledTableSnapshotHandler.prepare(EnabledTableSnapshotHandler.java:47) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.snapshotTable(SnapshotManager.java:659) ~[classes/:?]
	... 4 more

```

Eventually it is `SnapshotCreationException` that gets thrown.

Reply:
With updated patch, it looks like this:

```
org.apache.hadoop.hbase.snapshot.SnapshotCreationException: Could not build snapshot handler
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.snapshotTable(SnapshotManager.java:678) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.snapshotEnabledTable(SnapshotManager.java:646) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.takeSnapshotInternal(SnapshotManager.java:762) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.takeSnapshot(SnapshotManager.java:707) ~[classes/:?]
	at org.apache.hadoop.hbase.master.procedure.TestSnapshotProcedureWithLockTimeout$1.run(TestSnapshotProcedureWithLockTimeout.java:105) ~[test-classes/:?]
Caused by: org.apache.hadoop.hbase.DoNotRetryIOException: Master lock could not be acquired
	at org.apache.hadoop.hbase.master.snapshot.TakeSnapshotHandler.prepare(TakeSnapshotHandler.java:165) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.EnabledTableSnapshotHandler.prepare(EnabledTableSnapshotHandler.java:61) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.EnabledTableSnapshotHandler.prepare(EnabledTableSnapshotHandler.java:47) ~[classes/:?]
	at org.apache.hadoop.hbase.master.snapshot.SnapshotManager.snapshotTable(SnapshotManager.java:659) ~[classes/:?]
	... 4 more
```

Reply:
done